python/libucxx/libucxx/load.py

```diff
@@ -28,27 +55,29 @@ def load_library():
         libucx.load_library()
         del libucx
 
-    # Dynamically load libucxx.so. Prefer a system library if one is present to
-    # avoid clobbering symbols that other packages might expect, but if no
-    # other library is present use the one in the wheel.
+    prefer_system_installation = (
+        os.getenv("RAPIDS_LIBUCXX_PREFER_SYSTEM_LIBRARY", "false").lower() != "false"
+    )
+
+    soname = "libucxx.so"
     libucxx_lib = None
-    try:
-        libucxx_lib = ctypes.CDLL("libucxx.so", ctypes.RTLD_GLOBAL)
-    except OSError:
-        # If neither of these directories contain the library, we assume we are in an
-        # environment where the C++ library is already installed somewhere else and the
-        # CMake build of the libucxx Python package was a no-op. Note that this approach
-        # won't work for real editable installs of the libucxx package, but that's not a
-        # use case I think we need to support. scikit-build-core has limited support for
-        # importlib.resources so there isn't a clean way to support that case yet.
-        for lib_dir in ("lib", "lib64"):
-            if os.path.isfile(
-                lib := os.path.join(os.path.dirname(__file__), lib_dir, "libucxx.so")
-            ):
-                libucxx_lib = ctypes.CDLL(lib, ctypes.RTLD_GLOBAL)
-                break
+    if prefer_system_installation:
+        # Prefer a system library if one is present to
+        # avoid clobbering symbols that other packages might expect, but if no
+        # other library is present use the one in the wheel.
+        try:
+            libucxx_lib = _load_system_installation(soname)
+        except OSError:
+            libucxx_lib = _load_wheel_installation(soname)
+    else:
+        # Prefer the libraries bundled in this package. If they aren't found
+        # (which might be the case in builds where the library was prebuilt
+        # before packaging the wheel), look for a system installation.
+        libucxx_lib = _load_wheel_installation(soname)
+        if libucxx_lib is None:
+            libucxx_lib = _load_system_installation(soname)
 
     # The caller almost never needs to do anything with this library, but no
     # harm in offering the option since this object at least provides a handle
-    # to inspect where libucxx was loaded from.
+    # to inspect where the libucxx was loaded from.
     return libucxx_lib
```

Comment:
Note that this logic breaks the case of editable installs for developers in (for example) the devcontainer setup.

In that scenario, we don't have a wheel-provided DSO, and we don't have a "system" provided one. Instead, the cython files in the ucxx package have the rpath to the locally built DSO baked in, so ld.so happily finds it. However, here, since we're just asking ld.so to resolve a DSO name, we fail.

Reply:
Thanks, just pushed a commit with the changes we discussed offline and seen in https://github.com/rapidsai/kvikio/pull/553. That should address that.